tools/syz-trace2syz/parser/intermediate_types.go

```diff
@@ -0,0 +1,531 @@
+type Field struct {
+	Key string
+	Val IrType
+}
+
+func newField(key string, val IrType) *Field {
+	return &Field{Key: key, Val: val}
+}
+
+// Name implements IrType Name()
+func (f *Field) Name() string {
+	return "Field Type"
+}
+
+func (f *Field) String() string {
+	return f.Val.String()
+}
+
+// Expression represents Ints, Flags, Arithmetic expressions
+type Expression interface {
+	IrType
+	Eval(*prog.Target) uint64
+}
+
+type expressionCommon struct {
+}
+
+// Name implements IrType Name()
+func (e *expressionCommon) Name() string {
+	return exprTypeName
+}
+
+type binOp struct {
+	expressionCommon
+	leftOp  Expression
+	op      operation
+	rightOp Expression
+}
+
+func newBinop(leftOperand, rightOperand IrType, Op operation) *binOp {
+	return &binOp{leftOp: leftOperand.(Expression), rightOp: rightOperand.(Expression), op: Op}
+}
+
+// Eval implements Expression's Eval()
+func (b *binOp) Eval(target *prog.Target) uint64 {
+	op1Eval := b.leftOp.Eval(target)
+	op2Eval := b.rightOp.Eval(target)
+	switch b.op {
+	case andOp:
+		return op1Eval & op2Eval
+	case orOp:
+		return op1Eval | op2Eval
+	case xorOp:
+		return op1Eval ^ op2Eval
+	case lshiftOp:
+		return op1Eval << op2Eval
+	case rshiftOp:
+		return op1Eval >> op2Eval
+	case timesOp:
+		return op1Eval * op2Eval
+	case minusOp:
+		return op1Eval - op2Eval
+	case plusOp:
+		return op1Eval + op2Eval
+	default:
+		log.Fatalf("Unable to handle op: %d", b.op)
+		return 0
+	}
+}
+
+// String implements IrType String()
+func (b *binOp) String() string {
+	return fmt.Sprintf("op1: %s op2: %s, operand: %v\n", b.leftOp.String(), b.rightOp.String(), b.op)
+}
+
+type unOp struct {
+	expressionCommon
+	op      operation
+	operand Expression
+}
+
+func newUnop(operand IrType, op operation) *unOp {
+	return &unOp{op: op, operand: operand.(Expression)}
+}
+
+// Eval implements Expression's Eval()
+func (u *unOp) Eval(target *prog.Target) uint64 {
+	opEval := u.operand.Eval(target)
+	switch u.op {
+	case onescompOp:
+		return ^opEval
+	case negOp:
+		return -opEval
+	default:
+		log.Fatalf("Unsupported Unop Op: %d", u.op)
+	}
+	return 0
+}
+
+// String implements IrType String()
+func (u *unOp) String() string {
+	return fmt.Sprintf("op1: %v operand: %v\n", u.operand, u.op)
+}
+
+// DynamicType represents cases where strace shows the before system call and after
+// E.g. <... getsockname resumed> {sa_family=AF_INET6,..., sin6_scope_id=0}, [112->28]
+type DynamicType struct {
+	BeforeCall Expression
+	AfterCall  Expression
+}
+
+func newDynamicType(before, after IrType) *DynamicType {
+	return &DynamicType{BeforeCall: before.(Expression), AfterCall: after.(Expression)}
+}
+
+// String implements IrType String()
+func (d *DynamicType) String() string {
+	return d.BeforeCall.String()
+}
+
+// Name implements IrType Name()
+func (d *DynamicType) Name() string {
+	return "Dynamic Type"
+}
+
+// Garbage type used to parse content inside of ending parenthesis from strace
```

Comment:
It is possible. I think we can create an epsilon production instead of actually producing structures. I wasn't sure if there was something useful in the parenthesis. So far nothing interesting has come up so I will try and skip it.